### ISSUE TITLE
Remove blocking IO inside async code

### ISSUE THREAD

**Issue by justinmoon** [fedimint/fedimint#1528]
AFAIK we currently block during async writes do at least one of our database implementations. IIRC, one blocking IO operation could stall the whole async runtime?

@dpc can explain more.

**Comment by dpc** [comment]
Doing blocking IO in async code can lead to the whole tokio executor thread getting stack for some time, not letting (some) other futures/tasks making progress.

I think @m1sterc001guy is aware of this. We might even have an issue for it already.

**Comment by m1sterc001guy** [comment]
Yep, started on working on this today. This issue was also mentioned [here](https://github.com/fedimint/fedimint/issues/1322) but it's good to have its own issue to track.

**Comment by m1sterc001guy** [comment]
Rocksdb just merged support for enabling async IO in their C API: https://github.com/facebook/rocksdb/commit/ff04fb154bd74fd0681baa83b478095207e2719d

The rust-rocksdb library is linking against a version of rocksdb that does not have this API available yet though. For a proof of concept, I forked the repo, updated rocksdb to master, built from source, and exposed the API in rust. https://github.com/m1sterc001guy/rust-rocksdb/tree/async_io

This is a (hacky) example of how to consume it in fedimint. https://github.com/m1sterc001guy/fedimint/tree/async_rocksdb

I think we should wait for a new rocksdb release, try to upstream the rust changes, and use this so that our reads don't block the CPU thread.

> AFAIK we currently block during async writes do at least one of our database implementations. IIRC, one blocking IO operation could stall the whole async runtime?

Small nit, but writes don't block in Rocksdb. They always write to memory then do compaction in the background. If it can't write to memory because the buffer is full, the write will fail. If this happens, ideally we should backoff for a little bit and re-try (more discussion here: https://github.com/fedimint/fedimint/discussions/1370)

**Comment by m1sterc001guy** [comment]
Best explanation of what rocksdb async IO actually does: http://rocksdb.org/blog/2022/10/07/asynchronous-io-in-rocksdb.html

**Comment by dpc** [comment]
@m1sterc001guy I might be reading it wrong but their async and Rust's async might be two completely different things.

**Comment by m1sterc001guy** [comment]
Hmm ok agreed on that. The terminology is definitely misleading :( I think we might still benefit from this but it looks like we'll need to implement our own mechanism for unblocking the thread that does the read.

**Comment by dpc** [comment]
>  we'll need to implement our own mechanism for unblocking the thread that does the read.


Probably we'll have to run the backend as a threadpool, with a receiving channel for starting transactions. The outside code will send "start transaction" with a callback channel which will get picked by one of the threads in a backing threadpool, which will created a new "transaction command channel" and send the sender as a response. Then the backing thread will loop on transaction commands and execute them, sending responses back (again - via callbacks), until final "commit" command, which will make the thread return to polling on the "start transaction" channel again.

Quite a bit of infrastructure, but should be rather straightforward.

**Comment by justinmoon** [comment]
We do a similar thing to the threadpool idea for bitcoin RPC

DPC posted this, too https://www.reddit.com/r/rust/comments/10pf7m8/comment/j6kxeui/?context=3
